### PR TITLE
Extend Candid/Macaroon support

### PR DESCRIPTION
**Pull request by monstermunchkin**

### CODE REVIEW

File: client/connection.go [hunk 1]
```diff
@@ -159,7 +165,7 @@ func ConnectSimpleStreams(url string, args *ConnectionArgs) (ImageServer, error)
 }
 
 // Internal function called by ConnectLXD and ConnectPublicLXD
-func httpsLXD(url string, args *ConnectionArgs) (ContainerServer, error) {
+func httpsLXD(host string, args *ConnectionArgs) (ContainerServer, error) {
```
Comment (stgraber) [inline client/connection.go]: We've traditionally been using `uri` when the `url` package is used to avoid masking it.

File: lxc/main.go [hunk 2]
```diff
@@ -316,6 +316,22 @@ func (c *cmdGlobal) PreRun(cmd *cobra.Command, args []string) error {
...
+
+	server, _, err := cs.GetServer()
+	if err != nil {
+		return err
+	}
+
+	candidURL := server.Writable().Config["candid.api.url"]
+	if candidURL != nil {
+		c.conf.CandidURL = candidURL.(string)
+	}
```
Comment (stgraber) [inline lxc/main.go]: I don't see how this code can work. What if the user uses something other than the default remote, what if they do operations against multiple remotes?

Also in the case where the user is adding a new remote, they can't actually know the candid URL until after they've authenticated through candid, so I don't think this approach will work at all.

File: lxd/api_1.0.go [hunk 3]
```diff
@@ -378,11 +380,10 @@ func doApi10UpdateTriggers(d *Daemon, nodeChanged, clusterChanged map[string]str
 			fallthrough
 		case "maas.api.key":
 			maasChanged = true
+		case "candid.api.expiry":
```
Comment (stgraber) [inline lxd/api_1.0.go]: Shouldn't that also set candidChanged then?

**Comment by stgraber** [comment]
This branch should also take care of restricting the list of valid domains on the server side.

**Comment by stgraber** [comment]
- `candid.api.expiry` should be renamed to just `candid.expiry`
 - `candid.domains` needs to be added
 - A new API extension needs to be introduced to track the new config keys
 - The new config keys need to be documented in doc/server.md and added to the bash completion profile
 - We need to find a different way to handle the domain cookie thing, we can't require the client to know the URL of the candid server ahead of time, nor should we ever attempt to store that information since LXD can be reconfigured at any time

**Comment by stgraber** [comment]
For the domain cookie thing, I wonder if there's some way to set the cookie so that it is sent regardless of the domain, if so, that may be an option.

Otherwise, look for some way to hook into the bakery code so you can set the cookie after it gets the redirect from LXD, setting it for the right domain at that point.

File: lxd/api_1.0.go (hunk 3 above)
Comment (monstermunchkin) [reply lxd/api_1.0.go]: It does: `fallthrough`

Comment (stgraber) [reply lxd/api_1.0.go]: Oh right, missed that somehow.

**Comment by monstermunchkin** [comment]
Added docs, renamed config keys, and found a way around the cookie thing. I'm still trying to figure out how to implement the domain validation/restriction on the server side.

**Comment by stgraber** [comment]
@monstermunchkin Mind capitalizing the two "rename XYZ..." commits, the lack of consistency with the others bothers me :)

**Comment by stgraber** [comment]
Actually, those two should just go away and get squashed with whatever commit introduced the original code

File: client/connection.go [hunk 4]
```diff
@@ -159,7 +162,7 @@ func ConnectSimpleStreams(url string, args *ConnectionArgs) (ImageServer, error)
 }
 
 // Internal function called by ConnectLXD and ConnectPublicLXD
-func httpsLXD(url string, args *ConnectionArgs) (ContainerServer, error) {
+func httpsLXD(uri string, args *ConnectionArgs) (ContainerServer, error) {
```
Comment (stgraber) [inline client/connection.go]: This change doesn't really appear needed anymore.

File: client/connection.go [hunk 5]
```diff
@@ -168,7 +171,7 @@ func httpsLXD(url string, args *ConnectionArgs) (ContainerServer, error) {
 	// Initialize the client struct
 	server := ProtocolLXD{
 		httpCertificate:  args.TLSServerCert,
-		httpHost:         url,
+		httpHost:         uri,
```
Comment (stgraber) [inline client/connection.go]: This change doesn't really appear needed anymore.

File: doc/server.md [hunk 6]
```diff
@@ -12,6 +12,8 @@ Key                             | Type      | Default   | API extension
 :--                             | :---      | :------   | :------------            | :----------
 backups.compression\_algorithm  | string    | gzip      | backup\_compression      | Compression algorithm to use for new images (bzip2, gzip, lzma, xz or none)
 candid.api.url                  | string    | -         | candid\_authentication   | URL of the the external authentication endpoint using Candid
+candid.expiry                   | integer   | 3600      | candid\_config           | Candid macaroon expiry
```
Comment (stgraber) [inline doc/server.md]: Probably worth noting that it's in seconds in the description.

File: doc/server.md [hunk 7]
```diff
@@ -12,6 +12,8 @@ Key                             | Type      | Default   | API extension
...
+candid.domains                  | string    | -         | candid\_config           | Allowed Candid domains
```
Comment (stgraber) [inline doc/server.md]: Separator should probably be documented too (comma separator with the parser stripping spaces would be consistent with others)

File: shared/version/api.go [hunk 8]
```diff
@@ -122,6 +122,7 @@ var APIExtensions = []string{
 	"container_full",
 	"candid_authentication",
 	"backup_compression",
+	"candid_config",
```
Comment (stgraber) [inline shared/version/api.go]: This needs to be documented in doc/api-extensions.md

File: lxd/api_1.0.go [hunk 9]
```diff
@@ -417,5 +421,17 @@ func doApi10UpdateTriggers(d *Daemon, nodeChanged, clusterChanged map[string]str
 			return err
 		}
 	}
+
+	if candidChanged {
+		endpoint := clusterConfig.CandidEndpoint()
+		expiry := clusterConfig.CandidExpiry()
+		domains := clusterConfig.CandidDomains()
+
+		err := d.setupExternalAuthentication(endpoint, expiry, strings.Split(domains, ","))
```
Comment (stgraber) [inline lxd/api_1.0.go]: We should allow both "," and ", "

Comment (stgraber) [reply lxd/api_1.0.go]: Got this fixed locally, will push once the builders are done rebuilding their cache.

**Pull request merged**